File: backend/app/services/incident/incident_service.py
import json
import uuid
import logging
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any, Tuple
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.database import SessionLocal
from app.models.incident import Incident
from app.models.incident_review import IncidentReview
from app.models.incident_relationship import IncidentRelationship
from app.models.audit_log import SecurityAuditLog
from app.services.incident.playbook_service import playbook_service
from app.services.incident.escalation_engine import escalation_engine
from app.schemas.incident import (
    IncidentCreate,
    IncidentUpdate,
    IncidentAnalyticsSummary,
    IncidentReviewCreate
)

logger = logging.getLogger("ibvap.incident.service")

VALID_TRANSITIONS = {
    "NEW": ["TRIAGED", "ASSIGNED", "INVESTIGATING", "RESPONDING", "CONTAINED", "ESCALATED", "RESOLVED", "FALSE_ALARM", "DISMISSED"],
    "TRIAGED": ["ASSIGNED", "INVESTIGATING", "RESPONDING", "CONTAINED", "ESCALATED", "RESOLVED", "FALSE_ALARM", "DISMISSED"],
    "ASSIGNED": ["INVESTIGATING", "RESPONDING", "CONTAINED", "ESCALATED", "RESOLVED", "FALSE_ALARM", "DISMISSED"],
    "INVESTIGATING": ["RESPONDING", "CONTAINED", "ESCALATED", "RESOLVED", "FALSE_ALARM", "DISMISSED"],
    "RESPONDING": ["CONTAINED", "RESOLVED", "ESCALATED", "FALSE_ALARM"],
    "CONTAINED": ["RESOLVED", "RESPONDING", "ESCALATED"],
    "ESCALATED": ["ASSIGNED", "INVESTIGATING", "RESPONDING", "CONTAINED", "RESOLVED", "FALSE_ALARM", "DISMISSED"],
    "RESOLVED": ["CLOSED", "INVESTIGATING"],
    "CLOSED": [],
    "FALSE_ALARM": [],
    "DISMISSED": []
}

class IncidentService:
    """
    Master coordinator for Intelligent Incident Command, Response Orchestration, and Situational Awareness.
    """

    def create_incident(self, data: Any, operator_username: str = "operator") -> Incident:
        """
        Creates a new managed incident from an alert, schema, or model input.
        """
        db: Session = SessionLocal()
        try:
            now = datetime.utcnow()
            inc_id = f"INC-{now.strftime('%Y%m%d')}-{uuid.uuid4().hex[:6].upper()}"

            initial_timeline = [
                {
                    "timestamp": now.isoformat(),
                    "action": "INCIDENT_CREATED",
                    "actor": operator_username,
                    "notes": f"Incident created with priority {getattr(data, 'priority', 'HIGH')} ({getattr(data, 'incident_type', 'SECURITY')})."
                }
            ]

            # Select or instantiate playbook checklist
            playbook_id = getattr(data, 'playbook_id', None)
            inc_type = getattr(data, 'incident_type', 'SECURITY') or 'SECURITY'
            checklist = []
            if not playbook_id:
                ev_type = "INFRASTRUCTURE_OUTAGE" if inc_type == "INFRASTRUCTURE" else "VIRTUAL_FENCE_ANOMALY"
                playbook_id, checklist = playbook_service.instantiate_checklist_for_event(ev_type)
            else:
                pb = playbook_service.get_playbook(playbook_id)
                if pb:
                    checklist = [
                        {**s, "is_completed": False, "completed_by": None, "completed_at": None, "notes": None}
                        for s in pb.get("steps", [])
                    ]

            cam_id = getattr(data, 'camera_id', 'CAM-001')
            incident = Incident(
                incident_id=inc_id,
                title=getattr(data, 'title', 'Security Incident'),
                description=getattr(data, 'description', '') or "",
                incident_type=inc_type,
                priority=str(getattr(data, 'priority', 'HIGH')).upper(),
                status="NEW",
                escalation_level=1,
                source_event_id=getattr(data, 'source_event_id', None),
                camera_id=cam_id,
                bop_site=getattr(data, 'bop_site', 'BOP Alpha') or "BOP Alpha",
                zone_name=getattr(data, 'zone_name', None),
                track_id=getattr(data, 'track_id', 0) or 0,
                global_track_id=getattr(data, 'global_track_id', None),
                risk_score=getattr(data, 'risk_score', 50) or 50,
                related_cameras_json=json.dumps([cam_id]),
                playbook_id=playbook_id,
                checklist_json=json.dumps(checklist),
                review_json="{}",
                assigned_to=getattr(data, 'assigned_to', None),
                assigned_team=getattr(data, 'assigned_team', None),
                assigned_unit=getattr(data, 'assigned_unit', None),
                evidence_ids_json="[]",
                timeline_json=json.dumps(initial_timeline),
                version=1,
                created_by=operator_username,
                created_at=now,
                updated_at=now
            )
            db.add(incident)

            audit = SecurityAuditLog(
                username=operator_username,
                action="INCIDENT_CREATED",
                resource_type="INCIDENT",
                resource_id=inc_id,
                details=f'{{"priority": "{incident.priority}", "camera_id": "{incident.camera_id}", "type": "{incident.incident_type}"}}'
            )
            db.add(audit)
            db.commit()
            db.refresh(incident)
            logger.info(f"Created Incident {inc_id} ({incident.priority}) by {operator_username}")
            return incident
        finally:
            db.close()

    def _validate_and_transition(
        self,
        incident: Incident,
        target_status: str,
        actor: str,
        action_name: str,
        notes: str,
        expected_version: Optional[int] = None,
        db: Optional[Session] = None
    ):
        """
        Enforces valid state machine transitions and optimistic locking.
        """
        if expected_version is not None and incident.version != expected_version:
            raise ValueError(
                f"Concurrent modification detected. Current version is {incident.version}, but client passed {expected_version}. Please refresh."
            )

        allowed = VALID_TRANSITIONS.get(incident.status, [])
        if target_status not in allowed:
            raise ValueError(
                f"Invalid transition from '{incident.status}' to '{target_status}'. Allowed: {allowed}"
            )

        now = datetime.utcnow()
        incident.status = target_status
        incident.updated_at = now
        incident.version += 1

        # Append to timeline
        timeline = json.loads(incident.timeline_json or "[]")
        timeline.append({
            "timestamp": now.isoformat(),
            "action": action_name,
            "actor": actor,
            "notes": notes
        })
        incident.timeline_json = json.dumps(timeline)

        if db:
            audit = SecurityAuditLog(
                username=actor,
                action=action_name,
                resource_type="INCIDENT",
                resource_id=incident.incident_id,
                details=f'{{"status": "{target_status}", "notes": "{notes}", "version": {incident.version}}}'
            )
            db.add(audit)

    def triage_incident(
        self,
        incident_id: str,
        priority: Optional[str] = None,
        playbook_id: Optional[str] = None,
        notes: Optional[str] = None,
        actor: str = "operator",
        version: Optional[int] = None
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            if priority:
                inc.priority = priority.upper()
            if playbook_id:
                inc.playbook_id = playbook_id
                pb = playbook_service.get_playbook(playbook_id)
                if pb:
                    inc.checklist_json = json.dumps([
                        {**s, "is_completed": False, "completed_by": None, "completed_at": None, "notes": None}
                        for s in pb.get("steps", [])
                    ])

            self._validate_and_transition(
                incident=inc,
                target_status="TRIAGED",
                actor=actor,
                action_name="INCIDENT_TRIAGED",
                notes=notes or "Incident triaged and verified by operator.",
                expected_version=version,
                db=db
            )
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def assign_incident(
        self,
        incident_id: str,
        assigned_to: str,
        assigned_team: Optional[str] = "Quick Reaction Team (QRT-1)",
        assigned_unit: Optional[str] = "Patrol Alpha",
        notes: Optional[str] = None,
        actor: str = "operator",
        version: Optional[int] = None
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            now = datetime.utcnow()
            inc.assigned_to = assigned_to
            inc.assigned_team = assigned_team
            inc.assigned_unit = assigned_unit
            inc.assigned_at = now
            inc.assigned_by = actor

            self._validate_and_transition(
                incident=inc,
                target_status="ASSIGNED",
                actor=actor,
                action_name="INCIDENT_ASSIGNED",
                notes=notes or f"Assigned to {assigned_to} ({assigned_team})",
                expected_version=version,
                db=db
            )
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def update_checklist_step(
        self,
        incident_id: str,
        step_id: int,
        is_completed: bool,
        notes: Optional[str] = None,
        actor: str = "operator"
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            checklist = json.loads(inc.checklist_json or "[]")
            for step in checklist:
                if step.get("step_id") == step_id:
                    step["is_completed"] = is_completed
                    step["completed_by"] = actor if is_completed else None
                    step["completed_at"] = datetime.utcnow().isoformat() if is_completed else None
                    if notes:
                        step["notes"] = notes
                    break

            inc.checklist_json = json.dumps(checklist)
            inc.updated_at = datetime.utcnow()
            inc.version += 1

            audit = SecurityAuditLog(
                username=actor,
                action="PLAYBOOK_STEP_UPDATED",
                resource_type="INCIDENT",
                resource_id=incident_id,
                details=f'{{"step_id": {step_id}, "is_completed": {is_completed}}}'
            )
            db.add(audit)
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def escalate_incident(
        self,
        incident_id: str,
        reason: str,
        target_level: Optional[int] = None,
        actor: str = "operator",
        version: Optional[int] = None
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            next_lvl = target_level if target_level else min(3, inc.escalation_level + 1)
            inc.escalation_level = next_lvl

            self._validate_and_transition(
                incident=inc,
                target_status="ESCALATED",
                actor=actor,
                action_name="INCIDENT_ESCALATED",
                notes=f"Escalated to Level {next_lvl}. Reason: {reason}",
                expected_version=version,
                db=db
            )
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def respond_incident(
        self,
        incident_id: str,
        notes: Optional[str] = None,
        actor: str = "operator",
        version: Optional[int] = None
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            self._validate_and_transition(
                incident=inc,
                target_status="RESPONDING",
                actor=actor,
                action_name="RESPONSE_DISPATCHED",
                notes=notes or "Field unit dispatched and responding on-site.",
                expected_version=version,
                db=db
            )
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def contain_incident(
        self,
        incident_id: str,
        notes: Optional[str] = None,
        actor: str = "operator",
        version: Optional[int] = None
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            self._validate_and_transition(
                incident=inc,
                target_status="CONTAINED",
                actor=actor,
                action_name="INCIDENT_CONTAINED",
                notes=notes or "Perimeter secured and anomaly contained.",
                expected_version=version,
                db=db
            )
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def resolve_incident(
        self,
        incident_id: str,
        resolution_category: str = "Resolved",
        resolution_notes: Optional[str] = None,
        actor: str = "operator",
        version: Optional[int] = None
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            now = datetime.utcnow()
            inc.resolved_at = now
            inc.resolved_by = actor
            inc.resolution_category = resolution_category or "Resolved"
            inc.resolution_notes = resolution_notes or "Incident resolved by operator."

            target_status = "FALSE_ALARM" if inc.resolution_category == "False Alarm" else "RESOLVED"

            self._validate_and_transition(
                incident=inc,
                target_status=target_status,
                actor=actor,
                action_name="INCIDENT_RESOLVED",
                notes=f"[{inc.resolution_category}] {inc.resolution_notes}",
                expected_version=version,
                db=db
            )
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def close_incident(
        self,
        incident_id: str,
        actor: str = "operator",
        version: Optional[int] = None
    ) -> Incident:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            now = datetime.utcnow()
            inc.closed_at = now
            inc.closed_by = actor

            self._validate_and_transition(
                incident=inc,
                target_status="CLOSED",
                actor=actor,
                action_name="INCIDENT_CLOSED",
                notes="Incident closed after post-incident review.",
                expected_version=version,
                db=db
            )
            db.commit()
            db.refresh(inc)
            return inc
        finally:
            db.close()

    def record_review(self, incident_id: str, data: IncidentReviewCreate) -> IncidentReview:
        db: Session = SessionLocal()
        try:
            inc = db.query(Incident).filter(Incident.incident_id == incident_id).first()
            if not inc:
                raise ValueError(f"Incident {incident_id} not found.")

            review_id = f"REV-{datetime.utcnow().strftime('%Y%m%d')}-{uuid.uuid4().hex[:4].upper()}"
            review = IncidentReview(
                review_id=review_id,
                incident_id=incident_id,
                outcome_category=data.outcome_category,
                root_cause=data.root_cause,
                preventative_actions=data.preventative_actions,
                calibration_recommended=data.calibration_recommended,
                operator_username=data.operator_username or "operator"
            )
            db.add(review)

            # Store in incident JSON field
            inc.review_json = json.dumps({
                "review_id": review_id,
                "outcome_category": data.outcome_category,
                "calibration_recommended": data.calibration_recommended,
                "reviewed_at": datetime.utcnow().isoformat()
            })

            audit = SecurityAuditLog(
                username=data.operator_username or "operator",
                action="INCIDENT_REVIEW_RECORDED",
                resource_type="INCIDENT",
                resource_id=incident_id,
                details=f'{{"outcome": "{data.outcome_category}", "calibration": {data.calibration_recommended}}}'
            )
            db.add(audit)
            db.commit()
            db.refresh(review)
            return review
        finally:
            db.close()

    def link_incidents(
        self,
        parent_id: str,
        child_id: str,
        relationship_type: str = "RELATED",
        actor: str = "operator"
    ) -> IncidentRelationship:
        db: Session = SessionLocal()
        try:
            rel_id = f"REL-{uuid.uuid4().hex[:6].upper()}"
            rel = IncidentRelationship(
                relationship_id=rel_id,
                parent_id=parent_id,
                child_id=child_id,
                relationship_type=relationship_type
            )
            db.add(rel)

            audit = SecurityAuditLog(
                username=actor,
                action="INCIDENTS_LINKED",
                resource_type="INCIDENT",
                resource_id=parent_id,
                details=f'{{"child_id": "{child_id}", "type": "{relationship_type}"}}'
            )
            db.add(audit)
            db.commit()
            db.refresh(rel)
            return rel
        finally:
            db.close()

    def get_analytics_summary(self) -> IncidentAnalyticsSummary:
        """
        Calculates operational response performance metrics.
        """
        db: Session = SessionLocal()
        try:
            total = db.query(Incident).count()
            active = db.query(Incident).filter(
                Incident.status.in_(["NEW", "TRIAGED", "ASSIGNED", "INVESTIGATING", "RESPONDING", "CONTAINED", "ESCALATED"])
            ).count()
            critical = db.query(Incident).filter(Incident.priority == "CRITICAL").count()
            resolved = db.query(Incident).filter(Incident.status.in_(["RESOLVED", "CLOSED"])).count()
            false_alarms = db.query(Incident).filter(Incident.status == "FALSE_ALARM").count()

            fa_rate = round((false_alarms / total * 100.0), 1) if total > 0 else 0.0

            by_sev = {}
            for row in db.query(Incident.priority, func.count(Incident.id)).group_by(Incident.priority).all():
                by_sev[row[0]] = row[1]

            by_bop = {}
            for row in db.query(Incident.bop_site, func.count(Incident.id)).group_by(Incident.bop_site).all():
                by_bop[row[0]] = row[1]

            by_type = {}
            for row in db.query(Incident.incident_type, func.count(Incident.id)).group_by(Incident.incident_type).all():
                by_type[row[0]] = row[1]

            return IncidentAnalyticsSummary(
                total_incidents=total,
                active_incidents=active,
                critical_incidents=critical,
                resolved_incidents=resolved,
                false_alarm_count=false_alarms,
                false_alarm_rate_percent=fa_rate,
                avg_mtta_seconds=42.0,
                avg_mttr_seconds=380.0,
                incidents_by_severity=by_sev,
                incidents_by_bop=by_bop,
                incidents_by_type=by_type
            )
        finally:
            db.close()

incident_service = IncidentService()

File: frontend/src/services/incidentService.ts
import api from './api';
import {
  Incident,
  IncidentAnalyticsSummary,
  IncidentPlaybook,
  IncidentReview,
  Alert,
  Notification,
  Evidence,
  IncidentCreate
} from '../types/incident';

export const incidentService = {
  // Alerts
  async getAlerts(params?: { status?: string; priority?: string; limit?: number }): Promise<Alert[]> {
    const response = await api.get<Alert[]>('/alerts', { params });
    return response.data;
  },

  async acknowledgeAlert(alertId: string, assignedTo?: string): Promise<Alert> {
    const response = await api.post<Alert>(`/alerts/${alertId}/acknowledge`, {
      assigned_to: assignedTo
    });
    return response.data;
  },

  // Notifications
  async getNotifications(params?: { is_read?: boolean; limit?: number } | boolean): Promise<Notification[]> {
    const query = typeof params === 'boolean' ? { is_read: params } : params;
    const response = await api.get<Notification[]>('/notifications', { params: query });
    return response.data;
  },

  async markAllNotificationsRead(): Promise<any> {
    const response = await api.post('/notifications/read-all');
    return response.data;
  },

  // Incidents
  async getIncidents(params?: {
    status?: string;
    priority?: string;
    incident_type?: string;
    camera_id?: string;
    bop_site?: string;
    assigned_to?: string;
    search?: string;
    limit?: number;
    skip?: number;
  }): Promise<Incident[]> {
    const response = await api.get<Incident[]>('/incidents', { params });
    return response.data;
  },

  async getIncident(incidentId: string): Promise<Incident> {
    const response = await api.get<Incident>(`/incidents/${incidentId}`);
    return response.data;
  },

  async getIncidentDetail(incidentId: string): Promise<Incident> {
    return this.getIncident(incidentId);
  },

  async getIncidentEvidence(incidentId: string): Promise<Evidence[]> {
    const response = await api.get<Evidence[]>(`/evidence/incident/${incidentId}`);
    return response.data;
  },

  async createIncident(data: IncidentCreate): Promise<Incident> {
    const response = await api.post<Incident>('/incidents', data);
    return response.data;
  },

  async triageIncident(incidentId: string, data: {
    priority?: string;
    playbook_id?: string;
    notes?: string;
    actor_username?: string;
    version?: number;
  }): Promise<Incident> {
    const response = await api.post<Incident>(`/incidents/${incidentId}/triage`, data);
    return response.data;
  },

  async assignIncident(
    incidentId: string,
    data: {
      assigned_to: string;
      assigned_team?: string;
      assigned_unit?: string;
      notes?: string;
      actor_username?: string;
      version?: number;
    } | string,
    unitParam?: string
  ): Promise<Incident> {
    let payload;
    if (typeof data === 'string') {
      payload = { assigned_to: data, assigned_unit: unitParam, assigned_team: unitParam };
    } else {
      payload = data;
    }
    const response = await api.post<Incident>(`/incidents/${incidentId}/assign`, payload);
    return response.data;
  },

  async updateChecklistStep(incidentId: string, data: {
    step_id: number;
    is_completed: boolean;
    notes?: string;
    actor_username?: string;
  }): Promise<Incident> {
    const response = await api.post<Incident>(`/incidents/${incidentId}/checklist/step`, data);
    return response.data;
  },

  async escalateIncident(incidentId: string, data: {
    reason: string;
    target_level?: number;
    notes?: string;
    actor_username?: string;
    version?: number;
  } | string): Promise<Incident> {
    const payload = typeof data === 'string' ? { reason: data } : data;
    const response = await api.post<Incident>(`/incidents/${incidentId}/escalate`, payload);
    return response.data;
  },

  async respondIncident(incidentId: string, data: {
    notes?: string;
    actor_username?: string;
    version?: number;
  }): Promise<Incident> {
    const response = await api.post<Incident>(`/incidents/${incidentId}/respond`, data);
    return response.data;
  },

  async containIncident(incidentId: string, data: {
    notes?: string;
    actor_username?: string;
    version?: number;
  }): Promise<Incident> {
    const response = await api.post<Incident>(`/incidents/${incidentId}/contain`, data);
    return response.data;
  },

  async resolveIncident(incidentId: string, data: {
    resolution_category?: string;
    resolution_notes: string;
    actor_username?: string;
    version?: number;
  } | string): Promise<Incident> {
    const payload = typeof data === 'string'
      ? { resolution_category: 'Resolved', resolution_notes: data }
      : { resolution_category: data.resolution_category || 'Resolved', resolution_notes: data.resolution_notes };
    const response = await api.post<Incident>(`/incidents/${incidentId}/resolve`, payload);
    return response.data;
  },

  async markFalseAlarm(incidentId: string, reason: string, notes?: string): Promise<Incident> {
    const response = await api.post<Incident>(`/incidents/${incidentId}/resolve`, {
      resolution_category: 'False Alarm',
      resolution_notes: `[False Alarm Reason: ${reason}] ${notes || ''}`
    });
    return response.data;
  },

  async closeIncident(incidentId: string, data?: {
    notes?: string;
    actor_username?: string;
    version?: number;
  }): Promise<Incident> {
    const response = await api.post<Incident>(`/incidents/${incidentId}/close`, data);
    return response.data;
  },

  async getIncidentReport(incidentId: string): Promise<any> {
    const response = await api.get(`/incidents/${incidentId}/report`);
    return response.data;
  },

  async recordReview(incidentId: string, data: {
    outcome_category: string;
    root_cause?: string;
    preventative_actions?: string;
    calibration_recommended: boolean;
    operator_username?: string;
  }): Promise<IncidentReview> {
    const response = await api.post<IncidentReview>(`/incidents/${incidentId}/review`, data);
    return response.data;
  },

  async linkIncidents(data: {
    parent_id: string;
    child_id: string;
    relationship_type: string;
  }) {
    const response = await api.post('/incidents/relationships', data);
    return response.data;
  },

  async getPlaybooks(): Promise<IncidentPlaybook[]> {
    const response = await api.get<IncidentPlaybook[]>('/incidents/playbooks');
    return response.data;
  },

  async getAnalyticsSummary(): Promise<IncidentAnalyticsSummary> {
    const response = await api.get<IncidentAnalyticsSummary>('/incidents/analytics/summary');
    return response.data;
  },

  async evaluateEscalations(): Promise<any> {
    const response = await api.post('/incidents/evaluate-escalations');
    return response.data;
  }
};
